src/middlewared/middlewared/plugins/ldap.py

```diff
@@ -0,0 +1,641 @@
+            for sc in serverctrls:
+                if sc.controlType == SimplePagedResultsControl.controlType:
+                    cookie = sc.cookie
+                    if cookie:
+                        paged.cookie = cookie
+                        paged.size = self.pagesize
+
+                        break
+
+            if not cookie:
+                break
+
+            page += 1
+
+        return result
+
+    def get_samba_domains(self):
+        if not self._handle:
+            self._open()
+        filter = '(objectclass=sambaDomain)'
+        results = self._search(self.ldap['basedn'], ldap.SCOPE_SUBTREE, filter)
+        domains = []
+        if results:
+            for d in results:
+                domains.append(d[1]['sambaDomainName'][0].decode())
+
+        self._close()
+        return domains
+
+    def get_root_DSE(self):
+        """
+        root DSE query is defined in RFC4512 as a search operation
+        with an empty baseObject, scope of baseObject, and a filter of
+        "(objectClass=*)"
+        In theory this should be accessible with an anonymous bind. In practice,
+        it's better to use proper auth because configurations can vary wildly.
+        """
+        if not self._handle:
+            self._open()
+        filter = '(objectclass=*)'
+        results = self._search('', ldap.SCOPE_BASE, filter)
+        dse_objs = []
+        for r in results:
+            parsed_data = {}
+            for k, v in r[1].items():
+                v = list(i.decode() for i in v)
+                parsed_data.update({k: v})
+
+            dse_objs.append({
+                'dn': r[0],
+                'data': parsed_data
+            })
+        self._close()
+        return dse_objs
+
+
+class LDAPService(ConfigService):
+    class Config:
+        service = "ldap"
+        datastore = 'directoryservice.ldap'
+        datastore_extend = "ldap.ldap_extend"
+        datastore_prefix = "ldap_"
+
+    @private
+    async def ldap_extend(self, data):
+        data['hostname'] = data['hostname'].split()
+        return data
+
+    @private
+    async def ldap_compress(self, data):
+        data['hostname'] = ','.join(data['hostname'])
+        return data
+
+    @private
+    async def ldap_validate(self, ldap):
+        port = 636 if SSL(self.ldap['ssl']) == SSL.USESSL else 389
+        for h in ldap['hostname']:
+            self.middleware.call('ldap.port_is_listening', h, port, ldap['dns_timeout'])
+        self.middleware.call('ldap.validate_credentials')
+
+    @accepts(Dict(
+        'ldap_update',
+        List('hostname', required=True),
+        Str('basedn', required=True),
+        Str('binddn'),
+        Str('bindpw', private=True),
+        Bool('anonbind', default=False),
+        Str('usersuffix'),
+        Str('groupsuffix'),
+        Str('passwordsuffix'),
+        Str('machinesuffix'),
+        Str('sudosuffix'),
+        Str('ssl', default='off', enum=['off', 'on', 'start_tls']),
```

Comment:
@william-gr is it ok that we use lowercase constants here?

Reply:
Its not mandatory but I have asked if it could be easily changed to uppercase before (for all enums), seems its not trivial?

What do you think @anodos325 ? Too much effort?

Reply:
Not too much. I'll make the change here and in the AD plugin for consiste